probe/proc/cache.go

```diff
@@ -0,0 +1,123 @@
+package proc
+
+import (
+	"bytes"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	filesCacheLen        = 1024
+	filesCacheExpiration = 60 * time.Second
+)
+
+type filesCacheEntry struct {
+	sync.RWMutex
+	file     File
+	contents *bytes.Buffer
+	creation time.Time
+}
+
+func (fce *filesCacheEntry) Pinned() bool { return fce.contents != nil }
+func (fce *filesCacheEntry) ReadInto(buf *bytes.Buffer) error {
+	fce.Lock()
+	defer fce.Unlock()
+
+	if fce.Pinned() { // use the saved contents
+		buf.Write(fce.contents.Bytes())
+		return nil
+	}
+	return fce.file.ReadInto(buf)
+}
+
+// A cache for files handles
+type filesCache struct {
+	*lru.Cache
+	proc Dir
+	quit chan (struct{})
+}
+
+func newFilesCache(proc Dir) *filesCache {
+	evictionFunc := func(key, value interface{}) {
+		entry := value.(filesCacheEntry)
+		if entry.file != nil {
+			entry.file.Close()
+		}
+	}
+
+	cache, _ := lru.NewWithEvict(filesCacheLen, evictionFunc)
```

Comment:
To propagate this error was a mess, and `NewWithEvict` only raises this error when `len<0` (and we use a constant>0), so I don't know it's worth all the trouble...